### PR TITLE
Add dockerfile for pyenv and modify run_tests.py to include py35 and py36

### PR DESCRIPTION
**Pull request by thunderboltsid**
Replacement for [#6959 ](https://github.com/grpc/grpc/pull/6959). Fix for [#7110 ](https://github.com/grpc/grpc/issues/7110)
TIL: It's a bad idea to use master for submitting a PR.

@jtattermusch @nathanielmanistaatgoogle @kpayson64 please review.

### CODE REVIEW

File: tools/dockerfile/test/python_pyenv_x64/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,115 @@
+# Copyright 2015, Google Inc.
```
Comment (nathanielmanistaatgoogle) [inline tools/dockerfile/test/python_pyenv_x64/Dockerfile]: 2016's nearly half over.

**Comment by nathanielmanistaatgoogle** [comment]
Content looks mostly good.

[Limit your commit subject line to fifty characters](http://chris.beams.io/posts/git-commit/#seven-rules).

Add a lot more to your commit message. What will this change? Will this change the way that run_tests.py executes by default? Will this allow new values of flags to be specified?

File: tools/dockerfile/test/python_pyenv_x64/Dockerfile (hunk 1 above)
Comment (thunderboltsid) [inline tools/dockerfile/test/python_pyenv_x64/Dockerfile]: Fair enough. Fixed.

**Comment by nathanielmanistaatgoogle** [comment]
Please squash commits with each round of code review back-and-forth.

**Comment by jtattermusch** [comment]
@thunderboltsid   please don't add 4 assignees to a PR. I can give the PR a quick look, but if you assign 4 people that kind of means you want all 4 of us to do a full review and that's just a waste of resources.
You can do something like "@jtattermusch please take a quick look at the sanity of changes to Dockerfiles or so".

File: tools/dockerfile/test/python_pyenv_x64/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+FROM debian:jessie
```
Comment (jtattermusch) [inline tools/dockerfile/test/python_pyenv_x64/Dockerfile]: you should generate this docker file using a template under templates/  and then run `tools/buildgen/generate_projects.sh`.

File: tools/run_tests/run_tests.py [hunk 3]
```diff
@@ -434,6 +434,10 @@ def _get_tox_envs(self, compiler):
       return ('py27',)
     elif compiler == 'python3.4':
       return ('py34',)
+    elif compiler == 'python3.5':
```
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: you also need to actually use the dockerfile in you just added.

see `def dockerfile_dir(self):`

**Comment by nathanielmanistaatgoogle** [comment]
@jtattermusch: that was me who made the assignments; I'll provide more explicit guidance in the future.

In this case: @jtattermusch and one of @soltanmm, @kpayson64, and myself will do a full review, for a total of two full reviews.

**Comment by jtattermusch** [comment]
Added some comments and unassigned myself as I'll be unvavailable soon.

**Comment by nathanielmanistaatgoogle** [comment]
@nicolasnoble and one of @soltanmm, @kpayson64, and myself will do a full review, for a total of two full reviews. @nicolasnoble will review first and lead review.

**Comment by nicolasnoble** [comment]
@jtattermusch's comments need to be addressed.

File: tools/dockerfile/test/python_pyenv_x64/Dockerfile (hunk 2 above)
Comment (thunderboltsid) [inline tools/dockerfile/test/python_pyenv_x64/Dockerfile]: Running the script currently results in the following stack trace:
```
Traceback (most recent call last):
  File "src/boringssl/gen_build_yaml.py", line 163, in <module>
    generate_build_files.main([g])
  File "/Users/siddharthshukla/grpc_forks/thunderboltsid/grpc/third_party/boringssl/util/generate_build_files.py", line 468, in main
    stdout=err_data)
  File "/Users/siddharthshukla/.pyenv/versions/2.7-dev/lib/python2.7/subprocess.py", line 536, in check_call
    retcode = call(*popenargs, **kwargs)
  File "/Users/siddharthshukla/.pyenv/versions/2.7-dev/lib/python2.7/subprocess.py", line 523, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/Users/siddharthshukla/.pyenv/versions/2.7-dev/lib/python2.7/subprocess.py", line 711, in __init__
    errread, errwrite)
  File "/Users/siddharthshukla/.pyenv/versions/2.7-dev/lib/python2.7/subprocess.py", line 1343, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

Comment (soltanmm) [inline tools/dockerfile/test/python_pyenv_x64/Dockerfile]: Do you have `perl` and `go` (as in Go the language) installed?

**Comment by thunderboltsid** [comment]
@nicolasnoble Take a look now

**Comment by soltanmm** [comment]
Can anyone recall if this had a passing sanity test result for `generate_projects.sh` when we merged it?

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* One or more commits are not associated with a GitHub user. The email used in the Git commit must match a verified email address in a GitHub user profile. For more information, please see https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/

_Regards,
CLA GitHub bot_

**Pull request merged**